Update `minSdkVersion` (from `21` to `24`)

ParaskP7 opened the pull request:
This PR upgrades Login to `minSdkVersion` to `24`. This has been unblocked because:

- `WPAndroid` was already on `minSdkVersion = 24`, for some time now (see [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15135)).
- `WCAndroid` has been recently upgraded to `minSdkVersion = 24` as well (see [here](https://github.com/woocommerce/woocommerce-android/pull/7643)).

-----

As part of this `minSdkVersion = 24` upgrade the below changes were also applied in order to fix any additional warnings that this change brought-up:

Warnings Resolution List:

1. [Resolve obsolete sdk int lint warning for res values v23.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/20a7ea4f00f7d9c9f8308e8950f93efc220c00db)
2. [Remove unnecessary duplicate v27 theme login flow config.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/937f16449688cac7e9296f9cafea9a882abac3e9)

Dependency Config List:

1. [Rename kotlin ktx to androidx core version.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/b5aa5c8ee38743caba188494c88214dae3556b02)
2. [Extract all dependency version to root build gradle.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/7d132c4f5349a52f75ba9cc72c6acb2d9286ecf5)
3. [Add androidx prefix to app compat version.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/6b078f1e60fa069ef6a6c87e475906f739a1caa2)
4. [Move androidx app compat/core and dagger versions to main group.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/2507d767794c3896e8f7189c697485b2a2598752)

Dependency Update List:
1. [Update robolectric to 4.9.](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/commit/1a862eb14e7bb1c9d9932a1bc49fa4f3a6c7c235)

-----

PS: @ovitrif @atorresveiga I added you as the main reviewers, that is, in addition to the @wordpress-mobile/apps-infrastructure team itself, but randomly, since I just wanted someone from both, the `WordPress` and `Woo Commerce` teams, to sign-off on that change for WPAndroid and WCAndroid alike.

-----

### Testing instructions

- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough.
- However, if you want to be thorough about reviewing this change, you could:
    - Quickly smoke test, either the `WordPress` or `Woo Commerce` apps, with this version of `Login`, and see if it works as expected.
    - Try and install, either the `WordPress` or `Woo Commerce` apps, with this version of `Login`, on Android 5 (`API 21/22`) and Android 6 (`API 23`) devices and verify that you can't (see [API Levels](https://apilevels.com/)).

ParaskP7 committed:
Build: Extract dependency versions to root build gradle for all modules

This will now make sure that from now on all modules will be pointing to
the same dependency versions, making such upgrades much more
straightforward as that will apply to all modules. This will alleviate
the risk of forgetting to update a specific module, or set of modules
due to an oversight.

FYI: Although this repo has only one module at the moment, this doesn't
mean that another module wouldn't be created in the future, for example
an 'example' app module. As such, this commit is firstly done as a best
practice, then so that all repos start becoming more consistent with
each other and finally as a preparation step for new module creation.

ParaskP7 committed:
Build: Use double quotes instead of single quotes for dependencies

This is done because:
1. You get the below warning/error if you try to use a syntax like
'androidx.appcompat:appcompat:${version}', note the ${version}:
"It looks like you are trying to substitute a version variable, but
using single quotes ('). For Groovy string interpolation you must use
double quotes (")."
2. If this repo is ever to be migrated from Groovy to Kotlin Gradle DSL,
then this would become a prerequisite for the migration, that is, to
have double quotes (") everywhere.

ParaskP7 committed:
Deps: Rename kotlin ktx to androidx core version

The 'kotlin_ktx_version' naming is confusing and mind end up even being
misleading.

ParaskP7 committed:
Deps: Extract all dependency version to root build gradle

ParaskP7 committed:
Deps: Update robolectric to 4.9

This is done because otherwise the 'testDebugUnitTest' task fails the
test, when run on command line, with the below 'ClassNotFoundException'
error:

"java.lang.NoClassDefFoundError: android/content/ContentResolver
...
Caused by: java.lang.ClassNotFoundException:
couldn't load android.content.ContentResolver
...
Caused by: java.lang.IllegalArgumentException:
Unsupported class file major version 61"

Release Link: https://github.com/robolectric/robolectric/releases/tag/
robolectric-4.9

ParaskP7 committed:
Deps: Add androidx prefix to app compat version

ParaskP7 committed:
Deps: Move androidx app compat/core and dagger versions to main group

ParaskP7 committed:
Build: Extract common sdk versions to settings for all modules

This will now make sure that from now on all modules will be pointing to
the same 'min/compile/targetSdkVersion' versions, making such upgrades
much more straightforward as that will apply to all modules. This will
alleviate the risk of forgetting to update a specific module, or set of
modules due to an oversight.

FYI: Although this repo has only one module at the moment, this doesn't
mean that another module wouldn't be created in the future, for example
an 'example' app module. As such, this commit is firstly done as a best
practice, then so that all repos start becoming more consistent with
each other and finally as a preparation step for new module creation.

ParaskP7 committed:
Build: Update min sdk version to 24

A quick analysis showed that updating to 'minSdkVersion = 24' removed
one 'OldTargetAPI' warning and added just one 'ObsoleteSdkInt' new
warning.

All these newly created warnings will be dealt with in subsequent
commits. Also, any existing such warnings will be resolved too.

ParaskP7 committed:
Analysis: Resolve obsolete sdk int lint warning for res values v23

Warning Message: "This folder configuration (v23) is unnecessary;
minSdkVersion is 24. Merge all the resources in this folder into
values."

This 'res/values-v23' folder, its 'themes.xml' file and its
'Theme.LoginFlow' and 'Theme.LoginFlow.TransparentStatusBar' styles can
be moved into and replace the 'Theme.LoginFlow and empty
'Theme.LoginFlow.TransparentStatusBar' styles withing the 'themes.xml'
file in 'res/values', and thus now become the default, as the
'minSdkVersion' is now '24'.

ParaskP7 committed:
Analysis: Remove unnecessary duplicate v27 theme login flow config

Since this 20a7ea4f00f7d9c9f8308e8950f93efc220c00db change resolved the
obsolete sdk int lint warning for res values v23, this
'android:statusBarColor' and 'android:windowLightStatusBar' became the
default. As such, those can be now removed from teh v27 configuration.

ovitrif commented on WordPressLoginFlow/src/main/res/values-v27/themes.xml:
```diff
@@ -1,9 +1,6 @@
 <resources>
-
     <style name="Theme.LoginFlow" parent="Base.Theme.LoginFlow">
-        <item name="android:statusBarColor">?attr/colorSurface</item>
-        <item name="android:windowLightStatusBar">true</item>
```
I've noticed an issue with the `statusBar` color when testing WP-Android with the updates from this PR, on the login epilogue screen:

| Before | After |
| - | - |
| <img width="314" src="https://user-images.githubusercontent.com/4588074/201653884-bff6dd71-a7d1-4ee9-9ac6-781e557c07da.jpg"> | <img width="314" src="https://user-images.githubusercontent.com/4588074/201653877-fa60c962-0cd5-40de-9fd1-adc879952c44.jpg"> |

Not long ago I've encountered the same issue when working on the Jetpack Landing Screen redesign ([commit](https://github.com/wordpress-mobile/WordPress-Android/commit/df42df0407bbfbb400c6c29d74246d7af72bb533)), and by reviewing the documentation on [version specific styles](https://developer.android.com/develop/ui/views/theming/themes#Versions) I've learned that inheritance works a bit differently in styles, that in fact when we re-define a version-specific style it does not merge the properties from the main style with the properties from the version-specific one but actually overrides the whole style.

Because of the way it works, the `android:statusBarColor` and `android:windowLightStatusBar` would only be inherited in a "merging" fashion if they're declared in the `parent` style (`Base.Theme.LoginFlow`).

Though I see that the base theme only specifies colors, so maybe it's not meant to include the 2 we want to inherit 🤷 .

We could either:
1. Bring the 2 removed `statusBar` properties back to `v27` version specific style (which fixed the issue in my local test)
2. Create yet another base `parent` style from which to inherit the version-specific `Theme.LoginFlow` using the `parent` mechanism, that declares the `statusBar` properties only in the main theme file.

I hope this was clear, if not please let me know 🙇 .

ParaskP7 replied:
👋 @ovitrif !

> I've noticed an issue with the statusBar color when testing WP-Android with the updates from this PR, on the login epilogue screen:

Thank you so much for catching this early and before merging! 🙇 🙇 🙇 

> Not long ago I've encountered the same issue when working on the Jetpack Landing Screen redesign ([commit](https://github.com/wordpress-mobile/WordPress-Android/commit/df42df0407bbfbb400c6c29d74246d7af72bb533)), and by reviewing the documentation on [version specific styles](https://developer.android.com/develop/ui/views/theming/themes#Versions) I've learned that inheritance works a bit differently in styles, that in fact when we re-define a version-specific style it does not merge the properties from the main style with the properties from the version-specific one but actually overrides the whole style.

👍 

> Because of the way it works, the android:statusBarColor and android:windowLightStatusBar would only be inherited in a "merging" fashion if they're declared in the parent style (Base.Theme.LoginFlow).

Ahhh, this is true, I should have verifying this change myself but I assumed it would work by default, what could go wrong with style, right! I haven't worked with style long enough and next time I should be extra careful when trying to be a hero and optimize such styles instead of just letting it go... 🐈‍⬛ 

> Though I see that the base theme only specifies colors, so maybe it's not meant to include the 2 we want to inherit 🤷 .

🤷 

> We could either:
> 
> 1. Bring the 2 removed statusBar properties back to v27 version specific style (which fixed the issue in my local test)
> 2. Create yet another base parent style from which to inherit the version-specific Theme.LoginFlow using the parent mechanism, that declares the statusBar properties only in the main theme file.

I think the best and easier choice for now is for me to got with your option no.1 and revert my change there. 😊 

> I hope this was clear, if not please let me know 🙇 .

Ultra clear, once more, thank you! 🙇

ParaskP7 replied:
This is now done: 297fbd5340a6c89ea68ce670bd120cdbdc5c3777

ovitrif replied:
> I think the best and easier choice for now is for me to got with your option no.1 and revert my change there. 😊

Thank you for the prompt adaptation 🙇! I agree with this approach  🎉 🥇

ovitrif reviewed (commented):
Great work @ParaskP7 👏 , thank you for working on this 🙇 !!

I've tested the changes using WP-Android from `trunk` pointing to the local `Login-Flow` module updated to the version of this PR.

Overall the changes LGTM, I've left a comment for an issue I've noticed on the status bar color in WP-Android.

I understand from [this comment](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2563#discussion_r1020091519) that you may be reviewing & possibly change the approach with `gradle.ext...` versions in `settings.gradle`.

Please let me know when that's decided, as well as the resolution for the status bar color is handled, so I can give this another pass.

ParaskP7 committed:
Revert "Analysis: Remove unnecessary duplicate v27 theme login flow config"

This reverts commit 937f16449688cac7e9296f9cafea9a882abac3e9.

ParaskP7 commented:
👋 @ovitrif !

Thank you so much for reviewing, testing and catching [this](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/103#discussion_r1021459742) issue, you are a hero! 🥇 

> Overall the changes LGTM, I've left a comment for an issue I've noticed on the status bar color in WP-Android.

Replied [here](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/103#discussion_r1021540880). 🙇 🥇 🙏 

> I understand from https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2563#discussion_r1020091519 that you may be reviewing & possibly change the approach with gradle.ext... versions in settings.gradle.

Yes, exactly, thus, please let do the merging part, that is, when I figure-out how to best progress with the `gradle.ext` thingie... 🤔

> Please let me know when that's decided, as well as the resolution for the status bar color is handled, so I can give this another pass.

It is now ready for another pass, please take another look when possible, no rush! 💯

ovitrif reviewed (approved):
> It is now ready for another pass, please take another look when possible, no rush! 💯

Thank you Petros 🙇 

I've tested again the login flow in WP Android and the issue with the `statusBar` is resolved as expected 🥇 .

Approving this 🚀 🎉

ParaskP7 commented:
👋 @ovitrif !

> I've tested again the login flow in WP Android and the issue with the statusBar is resolved as expected 🥇 .

A BIG thank you Ovi! 🙇 🙇 🙇

atorresveiga reviewed (approved):
Thanks for working on this @ParaskP7!

I've tested the login flow in WCAndroid using the PR-commit dependency pattern and everything is working as expected.

LGTM! :shipit:

ParaskP7 commented:
Thank you so much for the review and testing @atorresveiga , you are awesome! 🙇 🥇 💯

ParaskP7 committed:
Build: Replace gradle ext with project ext for sdk versions declaration

For more info on that check PR comment and relevant discussion below:

https://github.com/wordpress-mobile/WordPress-FluxC-Android/
pull/2563#discussion_r1020079074
